### PR TITLE
warmstart blendsearch

### PR DESCRIPTION
**Pull request by sonichi**
Support warmstart with evaluated rewards
Increase test coverage
Only use define-by-run for hierarchical search space

### CODE REVIEW

File: flaml/tune/sample.py [hunk 1]
```diff
@@ -344,55 +339,6 @@ def domain_str(self):
         return f"{self.categories}"
 
 
-class Function(Domain):
```
Comment (qingyun-wu) [inline flaml/tune/sample.py:347]: why removing this?

File: test/test_training_log.py [hunk 2]
```diff
@@ -46,6 +46,7 @@ def test_training_log(self, path='test_training_log.log'):
             automl_settings["log_file_name"] = None
             automl_experiment.fit(X_train=X_train, y_train=y_train,
                                   **automl_settings)
+            automl_experiment._selected.update(None, 0)
```
Comment (qingyun-wu) [inline test/test_training_log.py:49]: This line is added intentionally to test the update function of SearchState? Does not look good.

Comment (sonichi) [reply test/test_training_log.py:49]: It's for improving the coverage. Any alternative way to improve coverage?

File: flaml/tune/sample.py (hunk 1 above)
Comment (sonichi) [reply flaml/tune/sample.py:347]: not tested and not used anywhere.

File: flaml/searcher/suggestion.py [hunk 3]
```diff
@@ -152,83 +143,6 @@ def on_trial_result(self, trial_id: str, result: Dict):
         """
         pass
 
-    def on_trial_complete(self,
```
[the remark this answers is not in the document]
Comment (sonichi) [reply flaml/searcher/suggestion.py:155]: not tested and not used anywhere.

File: flaml/scheduler/trial_scheduler.py [hunk 4]
```diff
@@ -31,127 +29,10 @@ class TrialScheduler:
...
-        raise NotImplementedError
-
-    def on_trial_error(self, trial_runner: "trial_runner.TrialRunner",
-                       trial: Trial):
-        """Notification for the error of trial.
-        This will only be called when the trial is in the RUNNING state."""
-
-        raise NotImplementedError
-
-    def on_trial_result(self, trial_runner: "trial_runner.TrialRunner",
```
[the remark this answers is not in the document]
Comment (sonichi) [reply flaml/scheduler/trial_scheduler.py:73]: It doesn't affect OnlineScheduler.

File: flaml/scheduler/trial_scheduler.py [hunk 5]
```diff
@@ -31,127 +29,10 @@ class TrialScheduler:
...
-
-    def on_trial_remove(self, trial_runner: "trial_runner.TrialRunner",
-                        trial: Trial):
-        """Called to remove trial.
-        This is called when the trial is in PAUSED or PENDING state. Otherwise,
-        call `on_trial_complete`."""
-
-        raise NotImplementedError
-
-    def choose_trial_to_run(
```
[the remark this answers is not in the document]
Comment (sonichi) [reply flaml/scheduler/trial_scheduler.py:98]: It doesn't affect OnlineScheduler.

File: test/test_training_log.py (hunk 2 above)
Comment (sonichi) [reply test/test_training_log.py]: Remember this is testing code.

**Review by qingyun-wu — APPROVED** [review]

File: flaml/automl.py [hunk 6]
```diff
@@ -456,30 +451,29 @@ def _validate_data(self, X_train_all, y_train_all, dataframe, label,
...
-                raise ValueError(
-                    "X_train_all must be a numpy array, a pandas dataframe, "
-                    "or Scipy sparse matrix.")
-            if not (isinstance(y_train_all, np.ndarray)
-                    or isinstance(y_train_all, pd.Series)):
-                raise ValueError(
-                    "y_train_all must be a numpy array or a pandas series.")
-            if X_train_all.size == 0 or y_train_all.size == 0:
-                raise ValueError("Input data must not be empty.")
+            assert (
```
Comment (ekzhu) [inline flaml/automl.py:454]: Generally `assert` is reserved for debugging or panic that must shutdown everything (see this [post](https://stackoverflow.com/questions/944592/best-practice-for-using-assert)). I think the old style is better -- it allows the user's program to catch and fix the error.

**Review by ekzhu — COMMENTED** [review]
Don't forget to update version number.

File: flaml/automl.py (hunk 6 above)
Comment (ekzhu) [reply flaml/automl.py:454]: An easy way to get around this is to replace `assert` with a custom function `check_value`, which raises a ValueError if the condition is false.

**Review by ekzhu — APPROVED** [review]

File: flaml/automl.py (hunk 6 above)
Comment (sonichi) [reply flaml/automl.py:453]: Agree to have a better approach in future. For the record, using assert is not totally unreasonable: https://code-examples.net/en/q/e69d0.

**Pull request merged**